Tests/ApolloCodegenTests/AnimalKingdomAPI/AnimalKingdomASTCreationTests.swift

```diff
@@ -0,0 +1,589 @@
+        .mock("PetDetails", type: GraphQLInterfaceType.mock("Pet")),
+        .mock("WarmBloodedDetails", type: GraphQLInterfaceType.mock("WarmBlooded")),
+      ]
+    )
+
+    // when
+    let actual = scope.mergedSelections
+
+    // then
+    expect(scope.type).to(equal(GraphQLInterfaceType.mock("WarmBlooded")))
+    expect(actual).to(shallowlyMatch(expected))
+  }
+
+  func test__mergedSelections_AllAnimalsQuery_AsCat__isCorrect() {
+    // given
+    let operation = Self.compilationResult.operations.first { $0.name == "AllAnimalsQuery" }
+    guard case let .field(allAnimals) = operation!.selectionSet.selections[0] else { fail(); return }
+    let allAnimalsScope = ASTSelectionSet(selectionSet: allAnimals.selectionSet!,
+                                          compilationResult: Self.compilationResult)
+    let scope = allAnimalsScope.children.values[2]
+
+    let expected = SortedSelections(
+      fields: [
+        .mock("isJellicle",
+              type: .nonNull(.named(GraphQLScalarType.boolean()))),
+        .mock("height",
+              type: .nonNull(.named(GraphQLObjectType.mock("Height")))),
+        .mock("species",
+              type: .nonNull(.named(GraphQLScalarType.string()))),
+        .mock("skinCovering",
+              type: .named(GraphQLEnumType.skinCovering())),
+        .mock("predators",
+              type: .nonNull(.list(.nonNull(.named(GraphQLInterfaceType.mock("Animal")))))),
+        .mock("humanName",
+              type: .named(GraphQLScalarType.string())),
+        .mock("favoriteToy",
+              type: .nonNull(.named(GraphQLScalarType.string()))),
+        .mock("owner",
+              type: .nonNull(.named(GraphQLObjectType.mock("Human")))),
+        .mock("bodyTemperature",
+              type: .nonNull(.named(GraphQLScalarType.integer()))),
+      ],
+      typeCases: [],
+      fragments: [
+        .mock("HeightInMeters", type: GraphQLObjectType.mock("Animal")),
+        .mock("WarmBloodedDetails", type: GraphQLInterfaceType.mock("WarmBlooded")),
+        .mock("PetDetails", type: GraphQLInterfaceType.mock("Pet")),
+      ]
+    )
+
+    // when
+    let actual = scope.mergedSelections
+
+    // then
+    expect(scope.type).to(equal(GraphQLObjectType.mock("Cat")))
+    expect(actual).to(shallowlyMatch(expected))
+  }
+
+#warning("TODO: This is the same as AllAnimal.AsPet.Height. Should we inherit that object instead?")
+  func test__mergedSelections_AllAnimalsQuery_AllAnimal_AsCat_Height__isCorrect() {
+    // given
+    let operation = Self.compilationResult.operations.first { $0.name == "AllAnimalsQuery" }
+    guard case let .field(allAnimals) = operation!.selectionSet.selections[0] else { fail(); return }
+    let allAnimalsScope = ASTSelectionSet(selectionSet: allAnimals.selectionSet!,
+                                          compilationResult: Self.compilationResult)
+    let height = allAnimalsScope
+      .children.values[2]
+      .mergedSelections
+      .fields.values[1]
```

Comment:
would be helpful to have some comments here about how we know these indexes

Reply:
These indexes are specifically the order that the fields/selections should be generated in the generated models. It's really very precise and the point is to ensure the models are generated exactly right with a deterministic ordering. I can add a comment.